spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java

```diff
@@ -386,7 +386,7 @@ private static SparkPartition toSparkPartition(CatalogTablePartition partition,
     Preconditions.checkArgument(serde.nonEmpty() || table.provider().nonEmpty(),
         "Partition format should be defined");
 
-    String uri = String.valueOf(locationUri.get());
+    URI uri = locationUri.get();
```

Comment:
Looks like the main problem is that `new Path(locationUri.get())` is not the same as `new Path(locationUri.get().toString())`?

Reply:
Yes, Like I wrote on the issue. The constructor of Path(string) assumes the string is decoded. The constructor of Path(Uri) has a URI so it knows it's encoded. That's why we can also fix this by reconverting back to a URI at the last moment (basically re asserting the encoding)

Reply:
```
scala> import java.net.URI
import java.net.URI

scala> import org.apache.hadoop.fs.Path
import org.apache.hadoop.fs.Path

scala> val uri = new URI("file:///has%20spaces")
uri: java.net.URI = file:///has%20spaces

scala> new Path(uri).toString
res4: String = file:/has spaces

scala> new Path(uri.toString).toString
res5: String = file:/has%20spaces
```

Reply:
Here what Spark does:

```
  /**
   * Convert URI to String.
   * Since URI.toString does not decode the uri, e.g. change '%25' to '%'.
   * Here we create a hadoop Path with the given URI, and rely on Path.toString
   * to decode the uri
   * @param uri the URI of the path
   * @return the String of the path
   */
  def URIToString(uri: URI): String = {
    new Path(uri).toString
  }

  /**
   * Convert String to URI.
   * Since new URI(string) does not encode string, e.g. change '%' to '%25'.
   * Here we create a hadoop Path with the given String, and rely on Path.toUri
   * to encode the string
   * @param str the String of the path
   * @return the URI of the path
   */
  def stringToURI(str: String): URI = {
    new Path(str).toUri
  }
```

Reply:
Okay, it sounds like we can keep this fix scoped to just these classes where we need to preserve the original URI instead of modifying it. That's a good rule for us to follow: if we have a URI, then use the URI constructor. If we have a String, use the String constructor.

Reply:
+1 for using the approach that @aokolnychyi pasted.

Reply:
Sounds good to me. I'd rather we didn't have 3 classes involved :) but If Spark does it this way at least we will be broken in similar ways if something doesn't work ;)

Reply:
Works at least for my little example

`scala> val uri = new URI("file:///has%20spaces")
uri: java.net.URI = file:///has%20spaces

scala> new Path(uri).toString
res0: String = file:/has spaces

scala> new Path(new Path(uri).toString)
res1: org.apache.hadoop.fs.Path = file:/has spaces`